database/transaction.go

```diff
@@ -340,19 +337,28 @@ func (tx *Transaction) ReIndex(indexName string) error {
 
 // ReIndexAll truncates and recreates all indexes of the database from scratch.
 func (tx *Transaction) ReIndexAll() error {
-	var indexes []string
+	it := tx.indexStore.st.Iterator(engine.IteratorOptions{})
+	itClose := it.Close
+	defer func() {
+		if itClose != nil {
+			itClose()
+		}
+	}()
 
-	it := tx.indexStore.st.NewIterator(engine.IteratorConfig{})
+	var indexes []string
 	for it.Seek(nil); it.Valid(); it.Next() {
 		indexes = append(indexes, string(it.Item().Key()))
 	}
-	err := it.Close()
-	if err != nil {
+	if err := it.Err(); err != nil {
 		return err
 	}
+	// Note that we can’t use deferred close because ReIndex needs
+	// an iterator but only one is allowed at a time.
+	itClose = nil
+	it.Close()
```

Comment:
Actually, this should be in a helper function so that we can defer it.Close() without this weird hack.